File: comm/tests/rp/clear_connections_spec.rs
// See comm/src/test/scala/coop/rchain/comm/rp/ClearConnectionsSpec.scala

use prost::bytes::Bytes;
use std::time::Duration;

use comm::rust::test_instances::{TransportLayerStub, NETWORK_ID};
use comm::rust::{
    errors::{timeout, CommError},
    peer_node::{Endpoint, NodeIdentifier, PeerNode},
    rp::{
        connect::{clear_connections, Connections, ConnectionsCell},
        rp_conf::{ClearConnectionsConf, RPConf},
    },
};

/// Helper function to create a peer with given name and default host/port
fn peer(name: &str) -> PeerNode {
    peer_with_host(name, "host")
}

/// Helper function to create a peer with given name and host
fn peer_with_host(name: &str, host: &str) -> PeerNode {
    let key = Bytes::from(name.as_bytes().to_vec());
    let id = NodeIdentifier { key };
    let endpoint = Endpoint::new(host.to_string(), 80, 80);
    PeerNode { id, endpoint }
}

/// Helper function to create a ConnectionsCell with given peers
fn mk_connections(peers: &[PeerNode]) -> ConnectionsCell {
    let connections_cell = ConnectionsCell::new();
    let connections = Connections::from_vec(peers.to_vec());
    connections_cell.flat_modify(|_| Ok(connections)).unwrap();
    connections_cell
}

/// Helper function to create RPConf for testing
fn conf(max_num_of_connections: usize, num_of_connections_pinged: Option<usize>) -> RPConf {
    RPConf {
        local: peer("src"),
        network_id: NETWORK_ID.to_string(),
        bootstrap: None,
        default_timeout: Duration::from_millis(1),
        max_num_of_connections,
        clear_connections: ClearConnectionsConf::new(num_of_connections_pinged.unwrap_or(5)),
    }
}

/// Always successful response function
fn always_success(
    _peer: &PeerNode,
    _protocol: &models::routing::Protocol,
) -> Result<(), CommError> {
    Ok(())
}

/// Always failing response function
fn always_fail(_peer: &PeerNode, _protocol: &models::routing::Protocol) -> Result<(), CommError> {
    Err(timeout())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_clear_connections_small_number_should_not_clear_any() {
        // given
        let connections = mk_connections(&[peer("A"), peer("B")]);
        let rp_conf = conf(5, None);
        let transport = TransportLayerStub::new();
        transport.set_responses(always_success);

        // when
        let cleared = clear_connections(&connections, &rp_conf, &transport)
            .await
            .unwrap();

        // then
        let final_connections = connections.read().unwrap();
        assert_eq!(final_connections.len(), 2);
        assert!(final_connections.as_slice().contains(&peer("A")));
        assert!(final_connections.as_slice().contains(&peer("B")));
        assert_eq!(cleared, 0);
    }

    #[tokio::test]
    async fn test_clear_connections_should_report_zero_cleared() {
        // given
        let connections = mk_connections(&[peer("A"), peer("B")]);
        let rp_conf = conf(5, None);
        let transport = TransportLayerStub::new();
        transport.set_responses(always_success);

        // when
        let cleared = clear_connections(&connections, &rp_conf, &transport)
            .await
            .unwrap();

        // then
        assert_eq!(cleared, 0);
    }

    #[tokio::test]
    async fn test_clear_connections_should_ping_first_few_nodes_with_heartbeat() {
        // given
        let connections = mk_connections(&[peer("A"), peer("B"), peer("C"), peer("D")]);
        let rp_conf = conf(5, Some(2));
        let transport = TransportLayerStub::new();
        transport.set_responses(always_success);

        // when
        clear_connections(&connections, &rp_conf, &transport)
            .await
            .unwrap();

        // then
        assert_eq!(transport.request_count(), 2);
        let requests = transport.get_all_requests();
        let peer_names: Vec<String> = requests
            .iter()
            .map(|req| String::from_utf8(req.peer.id.key.to_vec()).unwrap())
            .collect();
        assert!(peer_names.contains(&"A".to_string()));
        assert!(peer_names.contains(&"B".to_string()));
    }

    #[tokio::test]
    async fn test_clear_connections_should_remove_peers_that_did_not_respond() {
        // given
        let connections = mk_connections(&[peer("A"), peer("B"), peer("C"), peer("D")]);
        let rp_conf = conf(5, Some(2));
        let transport = TransportLayerStub::new();

        // Set responses: A fails, others succeed
        transport.set_responses(|peer, _protocol| {
            let peer_name = String::from_utf8(peer.id.key.to_vec()).unwrap();
            if peer_name == "A" {
                always_fail(peer, _protocol)
            } else {
                always_success(peer, _protocol)
            }
        });

        // when
        clear_connections(&connections, &rp_conf, &transport)
            .await
            .unwrap();

        // then
        let final_connections = connections.read().unwrap();
        assert_eq!(final_connections.len(), 3);
        assert!(!final_connections.as_slice().contains(&peer("A")));
        assert!(final_connections.as_slice().contains(&peer("B")));
        assert!(final_connections.as_slice().contains(&peer("C")));
        assert!(final_connections.as_slice().contains(&peer("D")));
    }

    #[tokio::test]
    async fn test_clear_connections_should_put_responding_peers_to_end_of_list() {
        // given
        let connections = mk_connections(&[peer("A"), peer("B"), peer("C"), peer("D")]);
        let rp_conf = conf(5, Some(3));
        let transport = TransportLayerStub::new();

        // Set responses: A fails, others succeed
        transport.set_responses(|peer, _protocol| {
            let peer_name = String::from_utf8(peer.id.key.to_vec()).unwrap();
            if peer_name == "A" {
                always_fail(peer, _protocol)
            } else {
                always_success(peer, _protocol)
            }
        });

        // when
        clear_connections(&connections, &rp_conf, &transport)
            .await
            .unwrap();

        // then
        let final_connections = connections.read().unwrap();
        assert_eq!(final_connections.len(), 3);

        // The order should be: D (not pinged), B, C (pinged and successful, moved to end)
        let connection_vec = final_connections.as_slice();
        assert_eq!(connection_vec[0], peer("D"));
        assert_eq!(connection_vec[1], peer("B"));
        assert_eq!(connection_vec[2], peer("C"));
    }

    #[tokio::test]
    async fn test_clear_connections_should_report_number_of_removed_connections() {
        // given
        let connections = mk_connections(&[peer("A"), peer("B"), peer("C"), peer("D")]);
        let rp_conf = conf(5, Some(3));
        let transport = TransportLayerStub::new();

        // Set responses: A fails, others succeed
        transport.set_responses(|peer, _protocol| {
            let peer_name = String::from_utf8(peer.id.key.to_vec()).unwrap();
            if peer_name == "A" {
                always_fail(peer, _protocol)
            } else {
                always_success(peer, _protocol)
            }
        });

        // when
        let cleared = clear_connections(&connections, &rp_conf, &transport)
            .await
            .unwrap();

        // then
        assert_eq!(cleared, 1);
    }
}

File: comm/src/test/scala/coop/rchain/comm/rp/ClearConnectionsSpec.scala
package coop.rchain.comm.rp

import scala.concurrent.duration._

import cats.{catsInstancesForId => _, _}

import coop.rchain.catscontrib._
import coop.rchain.catscontrib.ski._
import coop.rchain.catscontrib.effect.implicits._
import coop.rchain.comm._
import coop.rchain.comm.CommError._
import coop.rchain.comm.discovery._
import coop.rchain.comm.protocol.routing._
import coop.rchain.comm.rp.Connect._
import coop.rchain.comm.rp.ProtocolHelper._
import coop.rchain.metrics.Metrics
import coop.rchain.p2p.EffectsTestInstances.{LogicalTime, TransportLayerStub}
import coop.rchain.shared._

import org.scalatest._

class ClearConnectionsSpec
    extends FunSpec
    with Matchers
    with BeforeAndAfterEach
    with AppendedClues {

  import ScalaTestCats._

  val src: PeerNode                                 = peer("src")
  val networkId                                     = "test"
  implicit val transport                            = new TransportLayerStub[Id]
  implicit val log                                  = new Log.NOPLog[Id]
  implicit val metric                               = new Metrics.MetricsNOP[Id]
  implicit val time                                 = new LogicalTime[Id]
  implicit val kademliaStore: TrackingKademliaStore = new TrackingKademliaStore

  override def beforeEach(): Unit = {
    transport.reset()
    transport.setResponses(kp(alwaysSuccess))
    kademliaStore.removedKeys.clear()
  }

  describe("Node when called to clear connections") {
    describe(
      "if number of connections is smaller or equal to 2/3 of number of maximum connections allowed"
    ) {
      it("should not clear any of existing connections") {
        // given
        implicit val connections = mkConnections(peer("A"), peer("B"))
        implicit val rpconf      = conf(maxNumOfConnections = 5)
        // when
        Connect.clearConnections[Id]
        // then
        connections.read.size shouldBe 2
        connections.read should contain(peer("A"))
        connections.read should contain(peer("B"))
      }
      it("should report that 0 connections were cleared") {
        // given
        implicit val connections = mkConnections(peer("A"), peer("B"))
        implicit val rpconf      = conf(maxNumOfConnections = 5)
        // when
        val cleared = Connect.clearConnections[Id]
        // then
        cleared shouldBe 0
      }
    }

    describe("if number of connections is bigger then 2/3 of number of maximum connections allowed") {
      it("should ping first few nodes with heartbeat") {
        // given
        implicit val connections = mkConnections(peer("A"), peer("B"), peer("C"), peer("D"))
        implicit val rpconf      = conf(maxNumOfConnections = 5, numOfConnectionsPinged = 2)

        // when
        Connect.clearConnections[Id]
        // then
        transport.requests.size shouldBe 2
        transport.requests.map(_.peer) should contain(peer("A"))
        transport.requests.map(_.peer) should contain(peer("B"))
      }

      it("should remove connections of peers that did not respond to heartbeat") {
        // given
        implicit val connections = mkConnections(peer("A"), peer("B"), peer("C"), peer("D"))
        implicit val rpconf      = conf(maxNumOfConnections = 5, numOfConnectionsPinged = 2)
        transport.setResponses({
          case p if p == peer("A") => alwaysFail
          case _                   => alwaysSuccess
        })
        // when
        Connect.clearConnections[Id]
        // then
        connections.read.size shouldBe 3
        connections.read should not contain peer("A")
        connections.read should contain(peer("B"))
        connections.read should contain(peer("C"))
        connections.read should contain(peer("D"))
      }

      it("should put the peers that responded to heartbeat to the end of the list") {
        // given
        implicit val connections = mkConnections(peer("A"), peer("B"), peer("C"), peer("D"))
        implicit val rpconf      = conf(maxNumOfConnections = 5, numOfConnectionsPinged = 3)
        transport.setResponses({
          case p if p == peer("A") => alwaysFail
          case _                   => alwaysSuccess
        })
        // when
        Connect.clearConnections[Id]
        // then
        connections.read.size shouldBe 3
        connections.read shouldEqual List(peer("D"), peer("B"), peer("C"))
      }

      it("should report number of connections that were removed") {
        // given
        implicit val connections = mkConnections(peer("A"), peer("B"), peer("C"), peer("D"))
        implicit val rpconf      = conf(maxNumOfConnections = 5, numOfConnectionsPinged = 3)
        transport.setResponses({
          case p if p == peer("A") => alwaysFail
          case _                   => alwaysSuccess
        })
        // when
        val cleared = Connect.clearConnections[Id]
        // then
        cleared shouldBe 1
      }
    }

    describe("bootstrap peer protection") {
      it("should not remove bootstrap peer from KademliaStore when heartbeat fails") {
        // given
        implicit val connections = mkConnections(peer("A"), peer("B"), peer("C"), peer("D"))
        implicit val rpconf =
          conf(maxNumOfConnections = 5, numOfConnectionsPinged = 2, bootstrap = Some(peer("A")))
        transport.setResponses({
          case p if p == peer("A") => alwaysFail
          case _                   => alwaysSuccess
        })
        // when
        Connect.clearConnections[Id]
        // then
        kademliaStore.removedKeys should not contain peer("A").key
        // bootstrap is still removed from ConnectionsCell (TCP connection is broken)
        connections.read should not contain peer("A")
      }

      it("should still remove non-bootstrap peers from KademliaStore when heartbeat fails") {
        // given
        implicit val connections = mkConnections(peer("A"), peer("B"), peer("C"), peer("D"))
        implicit val rpconf =
          conf(maxNumOfConnections = 5, numOfConnectionsPinged = 2, bootstrap = Some(peer("BOOT")))
        transport.setResponses({
          case p if p == peer("A") => alwaysFail
          case _                   => alwaysSuccess
        })
        // when
        Connect.clearConnections[Id]
        // then
        kademliaStore.removedKeys should contain(peer("A").key)
      }
    }
  }

  private def peer(name: String, host: String = "host"): PeerNode =
    PeerNode(NodeIdentifier(name.getBytes), Endpoint(host, 80, 80))

  private def mkConnections(peers: PeerNode*): ConnectionsCell[Id] =
    Cell.id[Connections](peers.toList)

  private def conf(
      maxNumOfConnections: Int,
      numOfConnectionsPinged: Int = 5,
      bootstrap: Option[PeerNode] = None
  ): RPConfAsk[Id] =
    new ConstApplicativeAsk(
      RPConf(
        clearConnections = ClearConnectionsConf(numOfConnectionsPinged),
        defaultTimeout = 1.milli,
        local = peer("src"),
        networkId = networkId,
        bootstrap = bootstrap,
        maxNumOfConnections = maxNumOfConnections
      )
    )

  def alwaysSuccess: Protocol => CommErr[Unit] = kp(Right(()))

  def alwaysFail: Protocol => CommErr[Unit] = kp(Left(timeout))

  class TrackingKademliaStore extends KademliaStore[Id] {
    val removedKeys: scala.collection.mutable.ListBuffer[Seq[Byte]] =
      scala.collection.mutable.ListBuffer.empty

    def peers: Id[Seq[PeerNode]]                     = Seq.empty[PeerNode]
    def sparseness: Id[Seq[Int]]                     = Seq.empty[Int]
    def updateLastSeen(peerNode: PeerNode): Id[Unit] = ()
    def lookup(key: Seq[Byte]): Id[Seq[PeerNode]]    = Seq.empty[PeerNode]
    def find(key: Seq[Byte]): Id[Option[PeerNode]]   = None
    def remove(key: Seq[Byte]): Id[Unit]             = { removedKeys += key; () }
  }
}
